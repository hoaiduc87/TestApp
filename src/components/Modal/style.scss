@import 'assets/styles/mixins.scss';

.modal {
  position: fixed;
  z-index: $modal-index;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  &--visible {
    animation: fadeIn none 0.2s;
  }

  &--hidden {
    visibility: hidden;
    opacity: 0;
  }

  &--fullscreen {
    width: 100%;
    height: 100%;
    background-color: $black;

    .modal__dialog__close {
      position: absolute;
      right: 12px;
      top: 12px;
      padding: 4px;
      cursor: pointer;

      img {
        width: 32px;
      }
    }
  }

  &--small {
    width:  400px;
    box-shadow: 2px 2px 4px 2px $box-shadow-1;

    .modal__dialog__close {
      position: absolute;
      right: 10px;
      top: 10px;
      padding: 4px;
      cursor: pointer;

      img {
        width: 24px;
      }
    }
  }

  &__backdrop {
    position: absolute;
    z-index: $backdrop-index;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }

  &__dialog {
    position: absolute;
    z-index: $dialog-index;
  }
}

@keyframes fadeIn {
  0% {
      opacity: 0.5;
  }

  100% {
      opacity: 1;
  }
}

@include mobile {
  .modal--small {
    width: 80%;

    .modal__dialog__close {
      padding: 4px;
      right: 6px;
      top: 6px;
      cursor: pointer;

      img {
        width: 16px;
      }
    }
  }
}
