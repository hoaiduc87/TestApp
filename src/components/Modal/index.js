/* eslint-disable react-hooks/exhaustive-deps */
import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';

import CloseWhite from 'assets/images/close_white.svg';
import './style.scss';

function Modal({
  show,
  className,
  children,
  onClose,
}) {
  const handleKeyDown = (e) => {
    if (show && e.key === 'Escape') {
      onClose();
    }
  };

  useEffect(() => {
    if (show) {
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = 'auto';
    }

    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown)
    }
  }, [show]);

  return (
    <div className={`modal ${show ? 'modal--visible' : 'modal--hidden'}`}>
      <div className={`modal__dialog ${className}`}>
        <span className="modal__dialog__close" onClick={onClose}>
          <img src={CloseWhite} alt="close" />
        </span>
        {children || null}
      </div>
      <div className="modal__backdrop" onClick={onClose} />
    </div>
  )
}

Modal.propTypes = {
  show: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.any,
  onClose: PropTypes.func,
}

export default memo(Modal);
