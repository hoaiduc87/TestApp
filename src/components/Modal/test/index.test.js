import React from 'react';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';

import Modal from '..';

const props = {
  show: false,
  className: 'test',
};

describe('<Modal/>', () => {
  it('Should run and match snapshot', () => {
    const {
      container: { firstChild }
    } = render(<Modal {...props} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should execute onClose once click close button', () => {
    const onClose = jest.fn();
    const comp = mount(
      <Modal {...props} onClose={onClose}>
        <span>test</span>
      </Modal>
    );
    comp.setProps({ show: true });
    comp.find('.modal__dialog__close').first().simulate('click');
    expect(onClose).toHaveBeenCalled();
  });
})