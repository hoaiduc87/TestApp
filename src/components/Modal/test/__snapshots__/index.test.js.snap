// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> Should run and match snapshot 1`] = `
<div
  class="modal modal--hidden"
>
  <div
    class="modal__dialog test"
  >
    <span
      class="modal__dialog__close"
    >
      <img
        alt="close"
        src="close_white.svg"
      />
    </span>
  </div>
  <div
    class="modal__backdrop"
  />
</div>
`;
